re-process incomplete submissions

snopoke opened the pull request:
Not yet complete but opening for review

buddy @orangejenny 
cc @dimagi/scale-team

snopoke committed:
refactor form re-processor

snopoke committed:
initial test

snopoke committed:
[ci skip] spelling

orangejenny commented on corehq/form_processor/reprocess.py:
```diff
@@ -0,0 +1,146 @@
+                for case in cases:
+                    CaseAccessorSQL.save_case(case)
+
+                ledgers = _filter_already_processed_ledgers(form, stock_result.models_to_save)
```
Is `stock_result` guaranteed to exist? The `if stock_result` a few lines up makes me think it might be `None` and error here.

orangejenny replied:
nm, just realized this code is moved, not new

snopoke replied:
thanks, not sure why those checks were there since it's never None at this stage: 1052425

orangejenny commented:
I don't have much context in this area, but this looks reasonable to me.

calellowitz commented on corehq/form_processor/reprocess.py:
```diff
@@ -0,0 +1,146 @@
+        domain=form.domain, lock=True, deleted_ok=True, xforms=[form]
+    )
+    with cache as casedb:
+        case_stock_result = SubmissionPost.process_xforms_for_cases([form], casedb)
```
there appear to be a few cases where this can result in an exception, anything that would trigger `_handle_known_error` of `SubmissionPost` could create a stub that needs to be reprocessed but raise an exception here. additionally that should result in a form where `form.initial_processing_complete` remains `False` and the state becomes `error`

snopoke replied:
If there was an error raised when calling `process_xforms_for_cases` during the initial form processing then there wouldn't be a `stub`. But probably good to catch exceptions anyway.

calellowitz commented on corehq/form_processor/reprocess.py:
```diff
@@ -0,0 +1,146 @@
+def _reprocess_form(form):
+    # reset form state prior to processing
+    if should_use_sql_backend(form.domain):
+        form.state = XFormInstanceSQL.NORMAL
```
I think some errors and duplicate forms could end up with `UnfinishedSubmissionStubs` which this function does not account for

snopoke replied:
I don't think so - at least not in the current code. The `UnfinishedSubmissionStubs` are only created for 'NORMAL` forms

calellowitz reviewed (changes requested):
Added a few questions/comments

snopoke committed:
handle known errors

snopoke committed:
delete stub after successful processing

snopoke committed:
remove unecessary checks for stock_result

snopoke committed:
rename

snopoke committed:
re-raise erros if requested

snopoke committed:
test for case update

snopoke committed:
test for ledger creation

snopoke committed:
rebuild case and ledgers after re-processing

calellowitz reviewed (approved):
you are good on both comments. i got mixed up between calls to `self.save_processed_models` and `self.instance.save_processed_models`

snopoke committed:
only rebuild updated cases / ledgers

snopoke commented:
@calellowitz made some changes - most notably to rebuild cases / ledgers that get updated. I think this is ready to hook up to a task / mgmt command.

snopoke committed:
remove unused imports

snopoke committed:
move util

snopoke committed:
remove don't catch processing errors

We don't need to catch these when processing stubs
since the orignal form didn't raise any. If any are raised
then let them bubble up.

snopoke committed:
remove unused indexes

snopoke committed:
fix migration conflict

snopoke committed:
return result

snopoke committed:
lint

snopoke committed:
management command to reprocess manually

snopoke committed:
[ci skip] lint

snopoke commented:
@calellowitz This is ready to go. I want to do some manual testing before hooking it up to a task.

calellowitz reviewed (approved):
looks good to me. thanks for pushing this through

millerdev commented on corehq/form_processor/reprocess.py:
```diff
@@ -0,0 +1,206 @@
+        raise Exception('Form with ID {} not found'.format(form.form_id))
+
+    if not form.is_error:
+        raise Exception('Form was not an error form: {}={}'.format(form.form_id, form.doc_type))
```
Would it make sense to raise a custom subclass of `Exception` so it can be reliably caught without possibly masking bugs?

Also noticed several places below where `Exception` is raised.
